### ISSUE TITLE
[UX] Do not allow the "Administrator" role to be deleted (under certain conditions).

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1174]
You cannot delete the "Anonymous" or "Authenticated" roles, but you can delete the "Administrator" (same as any other user-created roles). Are we giving people the option to shoot themselves on the foot here? What is the rationale behind this?

![Screen Shot 2019-05-25 at 1 56 16 pm](https://user-images.githubusercontent.com/2423362/58364274-26874f80-7ef5-11e9-977a-d83573f9dc9b.png)

**Comment by sutibun** [comment]
Rationale? 

Keep the administrators on their toes ^\_^)

**Comment by MrHaroldA** [comment]
We don't use the Admin role here; so we'd like to be able to delete it.

**Comment by klonos** [comment]
Yes, but the admin role is the one assigned to user/1 and if you remove that role you basically lock yourself out of the installation. Right?

**Comment by docwilmot** [comment]
No, actually, the admin role a separate role that receives all permissions by default on install. For "Admins", if you need them. 

User/1 is different.

**Comment by klonos** [comment]
> User/1 is different.

Can you please elaborate on that? I have a live site I've noticed this on and although I'm "itchy" to delete the admin role as a test and see what actually happens, I'm afraid I might blow the site and have to start over.

**Comment by docwilmot** [comment]
User/1 is omnipotent; she always has all permissions, unless you specifically create a permissions rule in code which restricts this. E.g

`if(user->uid == 1) dont allow this`, or 

`if(user_doesnt have_this_custom_role_which_uid_1_doesnt) disallow that`.

Admins can still have roles delegated and removed in the UI.

**Comment by klonos** [comment]
Thanx for taking the time to explain @docwilmot. Is that for D7 too or just a Backdrop thing?

**Comment by docwilmot** [comment]
Both I believe.

**Comment by quicksketch** [comment]
If you delete a role which grants your current user the ability to assign/modify roles, you're going to get in deep water really quick. Because role permissions are checked on-the-fly, if you were a different admin user (not UID 1) and you deleted the admin role, you'd immediately get a 403 Access Denied page, because you'd no longer have permission to modify roles as soon as the form saved.

So maybe what we should do as a preventative measure is do a validation check that experiementally removes the role from the current user's `$user->roles` array, then checks access to modify roles with the about-to-be-deleted role removed. If the user is unable to modify roles after the role is deleted, throw a validation error and prevent the role from being deleted.

This makes it so a UID 1 could still delete the admin role if they wanted. Likewise this would allow the deleting of any role as long as it didn't lock the current user out. So if you were a non-UID-1 user with only the admin role, you wouldn't be able to delete it.

**Comment by MrHaroldA** [comment]
You must specify which role is the admin role, right? What if we don't allow an "active admin role" to be deleted?

So, if a role is set as admin role, that role cannot be deleted.

**Comment by quicksketch** [comment]
> You must specify which role is the admin role, right? What if we don't allow an "active admin role" to be deleted?

Yeah, that's a good option as well. Maybe we should do both things? One would prevent a user from accidentally locking themselves out. The other would prevent accidentally deleting a role that has a special ability.

**Comment by klonos** [comment]
I had a gut feeling that something was wrong about this. Just could not word it properly. Thank you guys for clarifying things. Both your ideas seem to make perfect sense.

**Comment by klonos** [comment]
> The other would prevent accidentally deleting a role that has a special ability.

So, if I get this right, we should NOT allow a role to be deleted if it is the last/only role that has the "Administer permissions" permission and/or(?) the "Administer users" permission. Right?

**Comment by quicksketch** [comment]
Yeah, that would be a side-effect, unless the user deleting the admin role were UID 1. In which case they always have "administer permissions", so they would be able to delete the admin role anyway.

**Comment by jenlampton** [comment]
As an update on this, there was a patch to Drupal 7 (also merged into Backdrop) that un-set the 'Admin role' setting when the admin role was deleted.  This allows people to add and delete roles willy-nilly, and after they are satisfied with roles, then they can return to set an admin role (or not).

One thing to note, Admin roles were new as of Drupal 7. Before D7 there were no admin roles and people managed alright with user 1 (and by creating their own roles.) I think there is a use-case for people not wanting/needing an admin role and returning to this D6 like state, so if we do add additional validation on the admin role when it is being deleted, we should be careful not to prevent it from being possible. 

I agree some additional warning would be an improvement :)

**Comment by MrHaroldA** [comment]
So IIUC, we should not allow the administrator role to be deleted, unless the user unsets the administrator role first?

Come to think of it; shouldn't the admin role assignment form be at admin/config/people/roles instead of admin/config/people/settings? We could add a help text there too about deleting the admin role.

**Comment by MrHaroldA** [comment]
Here's a mock-up showing the admin role fieldset, and the inability to delete admin roles directly.

![roles](https://cloud.githubusercontent.com/assets/695291/13425185/7f32fe06-dfa5-11e5-8c1a-97a9a8dcffe4.png)

**Comment by MrHaroldA** [comment]
@jenlampton or @quicksketch, is the mock-up anything you'd like? I can create a pull request for this after I polish some code ...

**Comment by klonos** [comment]
> If you delete a role which grants your current user the ability to assign/modify roles, you're going to get in deep water really quick. Because role permissions are checked on-the-fly, if you were a different admin user (not UID 1) and you deleted the admin role, you'd immediately get a 403 Access Denied page, because you'd no longer have permission to modify roles as soon as the form saved.
> 
> So maybe what we should do as a preventative measure is do a validation check that experiementally removes the role from the current user's `$user->roles` array, then checks access to modify roles with the about-to-be-deleted role removed. If the user is unable to modify roles after the role is deleted, throw a validation error and prevent the role from being deleted.
> 
> This makes it so a UID 1 could still delete the admin role if they wanted. Likewise this would allow the deleting of any role as long as it didn't lock the current user out. So if you were a non-UID-1 user with only the admin role, you wouldn't be able to delete it.

...

>> You must specify which role is the admin role, right? What if we don't allow an "active admin role" to be deleted?
>
>Yeah, that's a good option as well. Maybe we should do both things? One would prevent a user from accidentally locking themselves out. The other would prevent accidentally deleting a role that has a special ability.

I am interested in knocking this out, according to these recommendations.

...a couple of other things to consider:

- what happens with site where the admin role is already deleted?

- the "Administrator role" fieldset is collapsed by default. So when the admin role is deleted, the respective fieldset value is set to "disabled". This practically means that the feature which makes it so that admins are assigned all roles of newly-enabled modules is not in effect any longer. This is important, so we should expand that fieldset in that case.

**Comment by jenlampton** [comment]
> what happens with site where the admin role is already deleted?

This is the same as a site that does not have an "Administrator role" (but maybe no roles were ever deleted). 

In this case the checks would stop at "Is there an admin role?" and move on to the second validation check that experimentally removes the role from the current user's $user->roles array, then checks access to modify roles with the about-to-be-deleted role removed.

> when the admin role is deleted, the respective fieldset value is set to "disabled". 

I don't think that's the case. The only time the "Administrator role" option is disabled is when there is not a suitable role (when a site only has anonymous and authenticated). 

> This practically means that the feature which makes it so that admins are assigned all roles of newly-enabled modules is not in effect any longer. This is important, so we should expand that fieldset in that case.

How about we set a message if this happens, instead?

**Comment by BWPanda** [comment]
Renamed issue, as I was worried this issue was about preventing people deleting the Administrator role entirely, which I like to do occasionally.
